tree-sitter-stack-graphs/src/cli/match.rs

```diff
@@ -0,0 +1,161 @@
+/// Match file
+#[derive(Args)]
+pub struct MatchArgs {
+    /// Input file path.
+    #[clap(
+        value_name = "SOURCE_PATH",
+        required = true,
+        value_hint = ValueHint::AnyPath,
+        value_parser = ExistingPathBufValueParser,
+    )]
+    pub source_path: PathBuf,
+
+    /// Only match stanza on the given line.
+    #[clap(long, value_name = "LINE_NUMBER", short = 'S')]
+    pub stanza: Vec<usize>,
+}
+
+impl MatchArgs {
+    pub fn run(self, mut loader: Loader) -> anyhow::Result<()> {
+        let mut file_reader = FileReader::new();
+        let lc = match loader.load_for_file(&self.source_path, &mut file_reader, &NoCancellation)? {
+            Some(lc) => lc,
+            None => return Err(anyhow!("No stack graph language found")),
+        };
+        let source = file_reader.get(&self.source_path)?;
+        let tree = parse(lc.language, &self.source_path, source)?;
+        if self.stanza.is_empty() {
+            lc.sgl.tsg.try_visit_matches(&tree, source, true, |mat| {
+                print_matches(lc.sgl.tsg_path(), &self.source_path, source, mat)
+            })?;
+        } else {
+            for line in &self.stanza {
+                let stanza = lc
+                    .sgl
+                    .tsg
+                    .stanzas
+                    .iter()
+                    .find(|s| s.range.start.row <= line - 1 && line - 1 <= s.range.end.row)
+                    .ok_or_else(|| {
+                        anyhow!("No stanza on {}:{}", lc.sgl.tsg_path().display(), line)
+                    })?;
+                stanza.try_visit_matches(&tree, source, |mat| {
+                    print_matches(lc.sgl.tsg_path(), &self.source_path, source, mat)
+                })?;
+            }
+        }
+        Ok(())
+    }
+}
+
+fn print_matches(
+    tsg_path: &Path,
+    source_path: &Path,
+    source: &str,
+    mat: tree_sitter_graph::Match,
+) -> anyhow::Result<()> {
+    println!(
+        "{}: stanza query",
+        format!(
+            "{}:{}:{}",
+            tsg_path.display(),
+            mat.query_location().row + 1,
+            mat.query_location().column + 1
+        )
+        .bold(),
+    );
+    {
+        let full_capture = mat.full_capture();
+        print!("  matched ");
+        print_node(full_capture, true);
+        print_node_text(full_capture, source_path, source)?;
+        println!();
+    }
+    let width = mat
+        .capture_names()
+        .map(|n| n.len())
+        .max()
+        .unwrap_or_default();
+    if width == 0 {
+        return Ok(());
+    }
+    println!("  and captured");
+    for (name, quantifier, nodes) in mat.named_captures() {
+        let mut first = true;
+        for node in nodes {
+            if first {
+                first = false;
+                print!(
+                    "    @{}{}{} = ",
+                    name,
+                    quantifier_ch(quantifier),
+                    " ".repeat(width - name.len())
+                );
```

Comment:
Another nitpick: I would probably use `enumerate` and replace the `mut bool` with a check to see if the index is zero.

Reply:
I though about that as a solution too. It would make the code a bit neater, I think.